spike_main/spike.cc

```diff
@@ -123,31 +123,46 @@ bool sort_mem_region(const mem_cfg_t &a, const mem_cfg_t &b)
     return (a.base < b.base);
 }
 
-void merge_overlapping_memory_regions(std::vector<mem_cfg_t> &mems)
+static bool check_mem_overlap(const mem_cfg_t& L, const mem_cfg_t& R)
 {
-  // check the user specified memory regions and merge the overlapping or
-  // eliminate the containing parts
-  assert(!mems.empty());
+  const reg_t L_end   = L.base + L.size - 1;
+  const reg_t R_end   = R.base + R.size - 1;
+
+  return std::max(L.base, R.base) <= std::min(L_end, R_end);
+}
+
+static mem_cfg_t merge_mem_regions(const mem_cfg_t& L, const mem_cfg_t& R)
+{
+  // one can merge only intersecting regions
+  assert(check_mem_overlap(L, R));
+
+  const reg_t merged_base = std::min(L.base, R.base);
+  const reg_t merged_end_incl = std::max(L.base + L.size - 1, R.base + R.size - 1);
+  const reg_t merged_size = merged_end_incl - merged_base + 1;
+
+  return mem_cfg_t(merged_base, merged_size);
+}
+
+static void merge_overlapping_memory_regions(std::vector<mem_cfg_t> &mems)
+{
+  if (mems.empty())
+    return;
```

Comment:
This looks like a behavior change, since previously this was an assertion instead of a nop. Intentional?

Reply:
yes and no :).

"Yes" in the sense that  `merge_overlapping_memory_regions` does not need an assertion - the function itself  does not care of any invariant, and is only concerned about the merge.
"No" in the sense that probably I should preserve the assertion. To me it seems that if we need one - it's better to add it at the end `parse_mem_layout` function (right before we return the result).

Reply:
makes sense, thank you! addressed.

Reply:
The amended commit seems to put the assertion back into `merge_overlapping_memory_regions` which I thought we just agreed is not where it belongs? The old code had this assumption, but the new code does not, so the assertion doesn't really belong there anymore.

Reply:
> I thought we just agreed is not where it belongs?

My bad... Fixed.